### PR TITLE
Implements the committees/events search

### PR DESCRIPTION
**Pull request by romain-pierre**
Closes [EM-50](http://jira.sensiolabs.net/browse/EM-50) and [EM-51](http://jira.sensiolabs.net/browse/EM-51).

### CODE REVIEW

File: app/Resources/translations/messages.fr.yml [hunk 1]
```diff
@@ -43,3 +43,8 @@ committee.message_created: Votre message a bien été publié.
 committee.members_count: '1 membre|%count% membres'
 
 committee.event.creation.success: Le nouvel événement a bien été créé et publié sur la page du comité.
+
+#
+# Search
+#
+search.geocoding.exception: Nous ne sommes pas parvenu à reconnaitre la ville que vous avez saisi.
```
Comment (hhamon) [inline app/Resources/translations/messages.fr.yml]: parvenus
reconnaître

File: src/AppBundle/Search/SearchResultsProvider.php [hunk 2]
```diff
@@ -40,12 +39,32 @@ public function find(SearchParametersFilter $search): array
...
+                ->createNearbyQueryBuilder($coordinates)
+                ->orderBy('distance_between', 'asc')
+                ->andWhere($this->committee->getNearbyExpression().' < :distance_max')
+                ->setParameter('distance_max', $search->getRadius())
+            ;
+        } else {
+            $qb = $this->committee->createQueryBuilder('c');
+        }
+
+        if (!empty($search->getQuery())) {
```
Comment (HeahDude) [inline src/AppBundle/Search/SearchResultsProvider.php]: `if ($searchQuery = $search->getQuery())) {`, then reuse the var two lines below.

File: src/AppBundle/Search/SearchResultsProvider.php [hunk 3]
```diff
@@ -57,17 +76,35 @@ public function findCommittees(): array
...
+                ->andWhere($this->committee->getNearbyExpression().' < :distance_max')
+                ->setParameter('distance_max', $search->getRadius())
+            ;
+        } else {
+            $qb = $this->committee->createQueryBuilder('n');
+        }
 
-        /** @var CommitteeEvent */
-        $events = $this->event->findBy([], ['name' => 'asc'], 20);
+        if (!empty($search->getQuery())) {
```
Comment (HeahDude) [inline src/AppBundle/Search/SearchResultsProvider.php]: Same here.

File: app/Resources/translations/messages.fr.yml [hunk 4]
```diff
@@ -43,3 +43,8 @@ committee.message_created: Votre message a bien été publié.
 committee.members_count: '1 membre|%count% membres'
 
 committee.event.creation.success: Le nouvel événement a bien été créé et publié sur la page du comité.
+
+#
+# Search
+#
+search.geocoding.exception: Nous ne sommes pas parvenus à reconnaître la ville que vous avez saisi.
```
Comment (tgalopin) [inline app/Resources/translations/messages.fr.yml]: `saisie`

File: app/Resources/views/search/_committees.html.twig [hunk 5]
```diff
@@ -0,0 +1,9 @@
+<div style="float: left; width: 340px; height: 200px; background-color: #eee; margin: 10px; padding: 10px">
```
Comment (hhamon) [inline app/Resources/views/search/_committees.html.twig]: Avoid inline CSS style.

File: app/Resources/views/search/_events.html.twig [hunk 6]
```diff
@@ -0,0 +1,17 @@
+<div style="margin: 10px">
+    <img src="{{ path('map_url', { latitude: result.event.latitude, longitude: result.event.longitude }) }}" style="float: right; width: 200px">
+
+    <p class="text--medium">{{ result.event.name }}</p>
+    <p>
+        {{ result.event.beginAt|date('d F Y à h\\h') }}
```
Comment (hhamon) [inline app/Resources/views/search/_events.html.twig]: better to use a |locallizeddate filter.

File: app/Resources/views/search/index.html.twig [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+          {# this widget will be embedded in the ribbon text #}
+          <input name="c" value="{% if app.user.cityName is defined %}{{ app.user.cityName }}{% else %}{{ constant('AppBundle\\Search\\SearchParametersFilter::DEFAULT_CITY') }}{% endif %}" placeholder="Ville ou code postal">
+
+          <label><input type="radio" name="{{ constant('AppBundle\\Search\\SearchParametersFilter::PARAMETER_TYPE') }}" value="{{ constant('AppBundle\\Search\\SearchParametersFilter::TYPE_COMMITTEES') }}" checked="checked"> Comités</label>
+          <label><input type="radio" name="{{ constant('AppBundle\\Search\\SearchParametersFilter::PARAMETER_TYPE') }}" value="{{ constant('AppBundle\\Search\\SearchParametersFilter::TYPE_EVENTS') }}"> Evénements</label>
+      </div>
+      <div id="results"></div>
+  </div>
+
+  <script src="https://code.jquery.com/jquery-3.1.1.min.js"></script>
```
Comment (hhamon) [inline app/Resources/views/search/index.html.twig]: We don't rely on jQuery.

File: app/Resources/views/search/results.html.twig [hunk 8]
```diff
@@ -0,0 +1,13 @@
+{% for error in errors %}
+    <div class="error">
+      {{ error }}
+    </div>
+{% endfor %}
+
+{% for result in results %}
+    {{ include('search/_%s.html.twig'|format(search.type)) }}
```
Comment (hhamon) [inline app/Resources/views/search/results.html.twig]:
```twig
{{ include("search/_#{search.type}.html.twig") }}
```

File: src/AppBundle/DataFixtures/ORM/LoadCommitteeEventData.php [hunk 9]
```diff
@@ -14,33 +14,82 @@
...
+            'finish_at' => 'now',
+            'capacity' => 50,
+        ]);
+
+        $event3 = $committeeEventFactory->createFromArray([
+            'uuid' => self::COMMITTEE_EVENT_3_UUID,
+            'author' => $author7->getUuid(),
+            'committee' => $committee4,
+            'name' => 'Réunion de réflexion bellifontaine',
+            'category' => 'Réu',
```
Comment (hhamon) [inline src/AppBundle/DataFixtures/ORM/LoadCommitteeEventData.php]: Valid category code can be found in the `EventCategories` class.

File: src/AppBundle/Search/SearchParametersFilter.php [hunk 10]
```diff
@@ -41,29 +49,46 @@ class SearchParametersFilter
...
+     *
+     * @return SearchParametersFilter
      */
     public function handleRequest(Request $request)
     {
+        $this->setQuery((string) $request->query->get(self::PARAMETER_QUERY));
         $this->setType($request->query->getAlpha(self::PARAMETER_TYPE));
         $this->setRadius($request->query->getInt(self::PARAMETER_RADIUS));
+
+        if (null !== $request->query->get(self::PARAMETER_CITY)) {
```
Comment (hhamon) [inline src/AppBundle/Search/SearchParametersFilter.php]:
```php
if ($city = $request->query->get(self::PARAMETER_CITY)) {
    $this->setCity((string) $city);
}
```

File: app/Resources/views/search/_committees.html.twig (hunk 5 above)
Comment (romain-pierre) [inline app/Resources/views/search/_committees.html.twig]: Will be integrated, so I think this is not important for now.

File: app/Resources/views/search/_committees.html.twig [hunk 11]
```diff
@@ -0,0 +1,15 @@
+{% for result in results %}
+    <div style="float: left; width: 340px; height: 200px; background-color: #eee; margin: 10px; padding: 10px">
+        <p class="text--medium">{{ result.committee.name }}</p>
+        {% if result.committee.cityName is not empty %}
```
Comment (hhamon) [inline app/Resources/views/search/_committees.html.twig]: Can the city name be empty?

File: app/Resources/views/search/_committees.html.twig [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+            <p><strong>{{ result.committee.cityName }}</strong></p>
+        {% endif %}
+
+        <a href="{{ path('app_committee_show', { uuid: result.committee.uuid, slug: result.committee.slug }) }}" class="btn btn--blue">Je Participe</a>
+        <i class="fa fa-users"></i> {{ '1 participant|%count% participants'|transchoice(result.memberships) }}
```
Comment (hhamon) [inline app/Resources/views/search/_committees.html.twig]: Use the translation key instead here.

File: app/Resources/views/search/_events.html.twig [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+        <h1>{{ result.event.beginAt|localizeddate('none', 'none', 'fr_FR', null, 'd MMMM Y') }}</h1>
+    {% endif %}
+
+    <div style="margin: 10px">
+        <img src="{{ path('map_url', { latitude: result.event.latitude, longitude: result.event.longitude }) }}" style="float: right; width: 200px">
+
+        <p class="text--medium">{{ result.event.name }}</p>
+        <p>
+            {{ result.event.beginAt|localizeddate('none', 'none', 'fr_FR', null, 'd MMMM Y') }}
+            {% if result.event.cityName is not empty %}
```
Comment (hhamon) [inline app/Resources/views/search/_events.html.twig]: Can the city name be empty?

File: app/Resources/views/search/_events.html.twig [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+                &mdash; à {{ result.event.cityName }}
+            {% endif %}
+        </p>
+        <p>Organisé par : <strong>{{ result.event.committee.name }}</strong></p>
+
+        <a href="{{ path('app_committee_show_event', { committee_uuid: result.event.committee.uuid, slug: result.event.slug }) }}" class="btn btn--blue">Je Participe</a>
+        <i class="fa fa-users"></i> {{ '1 participant|%count% participants'|transchoice(result.attendees) }}
```
Comment (hhamon) [inline app/Resources/views/search/_events.html.twig]: Use the translation key instead.

File: app/Resources/views/search/_committees.html.twig (hunk 11 above)
Comment (romain-pierre) [inline app/Resources/views/search/_committees.html.twig]: http://confluence.sensiolabs.net/pages/viewpage.action?pageId=8455988&preview=/8455988/8455987/Rechercher_des_comite%CC%81s%20(1).png

File: src/AppBundle/Repository/NearbyTrait.php [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+     */
+    public function getNearbyExpression(): string
+    {
+        return '(6371 * acos(cos(radians(:latitude)) * cos(radians(n.postAddress.latitude))
+            * cos(radians(n.postAddress.longitude) - radians(:longitude)) + sin(radians(:latitude)) *
+            sin(radians(n.postAddress.latitude))))';
+    }
+
+    /**
+     * Calculates the distance (in Km) between the committee and the provided geographical
```
Comment (tgalopin) [inline src/AppBundle/Repository/NearbyTrait.php]: I suppose you use this also for events, so I wouldn't put the name of the entity in the comment.

Comment (hhamon) [inline src/AppBundle/Repository/NearbyTrait.php]: 👍

Comment (romain-pierre) [inline src/AppBundle/Repository/NearbyTrait.php]: Oops, miss after refactoring :)

File: app/Resources/views/search/_events.html.twig [hunk 16]
```diff
@@ -0,0 +1,22 @@
+{% for result in results %}
+    {% if date is not defined or date != result.event.beginAt|date('Ymd') %}
+        {% set date = result.event.beginAt|date('Ymd') %}
```
Comment (hhamon) [inline app/Resources/views/search/_events.html.twig]: Am I right to say this variable is not used?

File: app/Resources/views/search/_events.html.twig [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+    <div style="margin: 10px">
+        <img src="{{ path('map_url', { latitude: result.event.latitude, longitude: result.event.longitude }) }}" style="float: right; width: 200px">
+
+        <p class="text--medium">{{ result.event.name }}</p>
+        <p>
+            {{ result.event.beginAt|localizeddate('none', 'none', 'fr_FR', null, 'd MMMM Y') }} &mdash; à {{ result.event.cityName }}
+        </p>
+        <p>Organisé par : <strong>{{ result.event.committee.name }}</strong></p>
+
+        <a href="{{ path('app_committee_show_event', { committee_uuid: result.event.committee.uuid, slug: result.event.slug }) }}" class="btn btn--blue">Je Participe</a>
```
Comment (hhamon) [inline app/Resources/views/search/_events.html.twig]: The route name must now be `app_committee_attend_event`.

File: app/Resources/views/search/index.html.twig [hunk 18]
```diff
@@ -0,0 +1,111 @@
+{% extends 'base.html.twig' %}
+
+{% block title %}
+    Recherche {{ parent() }}
+{% endblock %}
+
+{% block banner '' %}
+
+{% block javascripts %}
+    <script type="text/javascript" src="https://maps.googleapis.com/maps/api/js?key={{ gmaps_api_key }}&libraries=places"></script>
```
Comment (hhamon) [inline app/Resources/views/search/index.html.twig]: `/&/&amp;/`

**Comment by hhamon** [comment]
Can you please rebase your PR and fixes conflicts + changes.

File: app/Resources/views/search/_events.html.twig (hunk 16 above)
Comment (romain-pierre) [inline app/Resources/views/search/_events.html.twig]: `date` is used in the `for` context.

Comment (hhamon) [inline app/Resources/views/search/_events.html.twig]: But where?

File: app/Resources/views/search/_committees.html.twig [hunk 19]
```diff
@@ -0,0 +1,9 @@
+{% for result in results %}
+    <div style="float: left; width: 340px; height: 200px; background-color: #eee; margin: 10px; padding: 10px">
+        <p class="text--medium">{{ result.committee.name }}</p>
+        <p><strong>{{ result.committee.cityName }}</strong></p>
+
+        <a href="{{ path('app_committee_show', { uuid: result.committee.uuid, slug: result.committee.slug }) }}" class="btn btn--blue">Je Participe</a>
```
Comment (hhamon) [inline app/Resources/views/search/_committees.html.twig]: committee_path()

**Pull request merged**